From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kanaka Raju Nayana <kanaka.raju.nayana@intel.com>
Date: Tue, 9 Jun 2026 07:23:51 +0530
Subject: drm/gem: preserve object layout on kernel 5.15

The v5.15 kernel's drm_gem_object_free() reads 'funcs' at a fixed compiled
offset. On v5.15 'funcs' sits after '_resv', but the v6.17 backport header
placed 'gpuva' before 'funcs', shifting the offset and causing the kernel
to read the wrong bytes as NULL.

Use the target-kernel gpuva probe to preserve the position of funcs in the
imported shared GEM object. Xe object initialization is no longer included.

Signed-off-by: Kanaka Raju Nayana <kanaka.raju.nayana@intel.com>
---

diff --git a/include/drm/drm_gem.h b/include/drm/drm_gem.h
index d3a7b43..e57f903 100644
--- a/include/drm/drm_gem.h
+++ b/include/drm/drm_gem.h
@@ -394,56 +394,77 @@ struct drm_gem_object {
 	 */
 	struct dma_resv _resv;
 
-	/**
-	 * @gpuva: Fields used by GPUVM to manage mappings pointing to this GEM object.
-	 *
-	 * When DRM_GPUVM_IMMEDIATE_MODE is set, this list is protected by the
-	 * mutex. Otherwise, the list is protected by the GEMs &dma_resv lock.
-	 *
-	 * Note that all entries in this list must agree on whether
-	 * DRM_GPUVM_IMMEDIATE_MODE is set.
-	 */
-	struct {
-		/**
-		 * @gpuva.list: list of GPUVM mappings attached to this GEM object.
-		 *
-		 * Drivers should lock list accesses with either the GEMs
-		 * &dma_resv lock (&drm_gem_object.resv) or the
-		 * &drm_gem_object.gpuva.lock mutex.
-		 */
-		struct list_head list;
-
-		/**
-		 * @gpuva.lock: lock protecting access to &drm_gem_object.gpuva.list
-		 * when DRM_GPUVM_IMMEDIATE_MODE is used.
-		 *
-		 * Only used when DRM_GPUVM_IMMEDIATE_MODE is set. It should be
-		 * safe to take this mutex during the fence signalling path, so
-		 * do not allocate memory while holding this lock. Otherwise,
-		 * the &dma_resv lock should be used.
-		 */
-#ifdef BPM_DRM_GEM_GPUVA_MUTEX_NOT_PRESENT
-#ifdef CONFIG_LOCKDEP
-#ifndef BPM_DRM_GEM_GPUVA_LOCK_DEP_MAP_NOT_PRESENT
-		struct lockdep_map *lock_dep_map;
-#endif
-#endif
-#else
-		struct mutex lock;
-#endif
-
-	} gpuva;
-
-	/**
-	 * @funcs:
-	 *
-	 * Optional GEM object functions. If this is set, it will be used instead of the
-	 * corresponding &drm_driver GEM callbacks.
-	 *
-	 * New drivers should use this.
-	 *
-	 */
-	const struct drm_gem_object_funcs *funcs;
+#ifndef BPM_DRM_GEM_OBJECT_GPUVA_MEMBER_NOT_PRESENT
+	/**
+	 * @gpuva: Fields used by GPUVM to manage mappings pointing to this GEM object.
+	 *
+	 * When DRM_GPUVM_IMMEDIATE_MODE is set, this list is protected by the
+	 * mutex. Otherwise, the list is protected by the GEMs &dma_resv lock.
+	 *
+	 * Note that all entries in this list must agree on whether
+	 * DRM_GPUVM_IMMEDIATE_MODE is set.
+	 */
+	struct {
+		/**
+		 * @gpuva.list: list of GPUVM mappings attached to this GEM object.
+		 *
+		 * Drivers should lock list accesses with either the GEMs
+		 * &dma_resv lock (&drm_gem_object.resv) or the
+		 * &drm_gem_object.gpuva.lock mutex.
+		 */
+		struct list_head list;
+
+		/**
+		 * @gpuva.lock: lock protecting access to &drm_gem_object.gpuva.list
+		 * when DRM_GPUVM_IMMEDIATE_MODE is used.
+		 *
+		 * Only used when DRM_GPUVM_IMMEDIATE_MODE is set. It should be
+		 * safe to take this mutex during the fence signalling path, so
+		 * do not allocate memory while holding this lock. Otherwise,
+		 * the &dma_resv lock should be used.
+		 */
+#ifdef BPM_DRM_GEM_GPUVA_MUTEX_NOT_PRESENT
+#ifdef CONFIG_LOCKDEP
+#ifndef BPM_DRM_GEM_GPUVA_LOCK_DEP_MAP_NOT_PRESENT
+		struct lockdep_map *lock_dep_map;
+#endif
+#endif
+#else
+		struct mutex lock;
+#endif
+
+	} gpuva;
+#endif /* !BPM_DRM_GEM_OBJECT_GPUVA_MEMBER_NOT_PRESENT */
+
+	/**
+	 * @funcs:
+	 *
+	 * Optional GEM object functions. If this is set, it will be used instead of the
+	 * corresponding &drm_driver GEM callbacks.
+	 *
+	 * New drivers should use this.
+	 *
+	 * NOTE: Field order is conditional to match kernel ABI.
+	 */
+	const struct drm_gem_object_funcs *funcs;
+
+#ifdef BPM_DRM_GEM_OBJECT_GPUVA_MEMBER_NOT_PRESENT
+	/**
+	 * @gpuva:
+	 *
+	 * Provides the list of GPU VAs attached to this GEM object.
+	 *
+	 * Drivers should lock list accesses with the GEMs &dma_resv lock
+	 * (&drm_gem_object.resv) or a custom lock if one is provided.
+	 */
+	struct {
+		struct list_head list;
+
+#ifdef CONFIG_LOCKDEP
+		struct lockdep_map *lock_dep_map;
+#endif
+	} gpuva;
+#endif /* BPM_DRM_GEM_OBJECT_GPUVA_MEMBER_NOT_PRESENT */
 
 	/**
 	 * @lru_node:
-- 
2.43.0
